### ISSUE TITLE
Provide async methods

### ISSUE THREAD

**Issue by mganss** [Genbox/VirusTotal.NET#15]
In order to stay DRY, make the current methods async, then rewrite the sync methods by calling the async methods and `Wait()`.

**Comment by Genbox** [comment]
Usually you have 2 reasons to implement async:

1. Scalability - you can fire multiple concurrent requests at the same time, and given the server that receives the requests is not overloaded or limited, you might get the results faster.
2. Latency - High latency operations you have to wait for improves with async, as you don't have to wait for them.

Now, this API implementation does not benefit from 1, as we are limited to 4 requests each minute. Calling the API 4 times in a row vs. 4 times at once does not really make a big difference. I would also like to keep the VT public API free, which means no spamming of requests. They also support batch requests, which means you should ask for 4 hashes at the same time, instead of 4 requests with only 1 hash, making async even less efficient. Note that if this API ever evolve to support the private part of the VT API (it might happen within the next months), it would probably make a lot more sense to implement async.

As for reason 2, the VT API is relatively low latency. If you ask for 25 hashes, it will respond within a second in most cases. Currently, this does not warrant an implementation of async methods.

I see async methods as a feature, but there has to be a reason to implement that feature. Currently, I see little reason to add more complexity (threading) to the API. You have to consider the many people out there that are new to the Tasks framework in .NET, and think it is weird that all methods return a Task object instead of just returning their results. For the moment, I think it is reasonable for you to do new Task(() => vt.Scan("...")).Start(); instead of asking everyone else to do vt.Scan().Wait();

While I'm closing this issue, I consider it a feature request for the future, and your request is taken into consideration.

**Comment by mganss** [comment]
It's about asynchronous IO, not about multithreading. Anytime network requests are involved you're needlessly blocking the caller for a huge amount of CPU cycles (billions). Also, "a second" is a subjective experience, if you're on EDGE it might be a different story than broadband etc.

WIth RestSharp async is easy, you just do `await ExecuteAsync()` instead of `Execute()`. You can then expose two sets of methods: one is synchronous just like now (e.g. `GetUrlReport()`) and the other is asynchronous (e.g. `GetUrlReportAsync()`). To keep it DRY, the synchronous variant can be a one-liner that calls the asynchronous variant, e.g. `return GetUrlReportAsync().Result;`  For people who want to use the synchronous variant nothing changes.

**Comment by Genbox** [comment]
"Anytime network requests are involved you're needlessly blocking the caller for a huge amount of CPU cycles (billions)"

That depends on how you wait. If you use reset events with WaitOne() as the underlying .NET framework does when waiting for a response, the CPU can effectively switch context to another thread, causing the waiting thread to use no CPU. If the methods were async, and you .Wait() on it instead, it is exactly the same waiting using reset events. In terms of performance, adding async would increase the number of threads, and therefore the number of wasted CPU cycles, it is however negligible.

Cluttering the API with sync and async methods does not make much sense to me. Again, you can easily call a synchronous method in an asynchronous fashion by creating a new thread, which is exactly what the async keyword does for you under the covers.

**Comment by mganss** [comment]
I didn't mean spinning (busy waiting) but rather blocking, i.e. the waiting thread can't do other stuff in the mean time. In a desktop application, your GUI will freeze.

Async/await [does not create a new thread](http://blog.stephencleary.com/2013/11/there-is-no-thread.html). See [also](http://stackoverflow.com/a/27265877/1970064).

**Comment by Genbox** [comment]
That is not true. The blog/post you refer to are simplifying the issue and sums it up into a resounding "no". However, there are missing a few important details. Async does create threads when it encounters something CPU intensive, but it is also heavily optimized and can execute directly in the callers thread (and context) if the job takes little time, thus saving the overhead of creating a thread and new context (using a thread pool). It also have an optimization on IO bound operations where it uses IO completion ports, which can be more efficient in the case where multiple IO operations are queued - usually we create a thread and then queue the IO job, but with async/await, it queues the job from the callers thread to minimize overhead.

That being said, the results are still the same: there is an overhead in using async none the less from the fact that is has to measure execution calls, save contexts, multiple marshal calls to win32 and more. This is NOT a reason to not use async in the library, it is just to counter the argument that billions of CPU cycles are lost.

Also note that HttpWebRequest is not an awaitable and since the operation takes time, it will spawn a thread for each call anyway.

**Comment by mganss** [comment]
These posts are only about IO bound tasks (like HTTP requests), obviously if you have a CPU bound task it has to run somewhere and if it can't be the callers thread which it can't block then it must be another thread. But in this context here we're only talking about network requests (IO bound). The async/await keywords themselves never create new threads.

I'm not saying the CPU cycles are lost in the sense that the calling thread is consuming them without doing any real work (busy waiting). In the synchronous case the calling thread is just blocked (does nothing) for the amount of time the request takes to complete. So, if a request takes 10 seconds, the calling thread is blocked for 10 seconds. If that calling thread is the UI thread, your GUI will freeze for 10 seconds.

HttpWebRequest uses the traditional Asynchronous Programming Model (APM) and RestSharp wraps it to expose a Task based API. No threads.

**Comment by Genbox** [comment]
You have to understand that I'm looking at it from a library point of view. Introducing new language features like async that depends on Task limits the portability of the library and force the library to be build with a newer version of .NET due to the TPL dependency. Hopefully this will change soon as .NET Core comes out and we can bundle the TPL dependency directly in the library.

There is also the issue with new users. Think back to when you first started C# development and you saw 100 methods/properties/fields in intellisense, and half of them you did not even understand how to call (async callbacks). I've always tried to keep it as simple as possible API wise and do all the complex stuff under the hood. In this case, the library is just a simple wrapper for the HTTP API with some methods for convenience.

95% of all projects don't care if you spawn another thread to do async work as the memory overhead is small(ish), and since this API is only targeting the VT public API where you have a limit of 4 request/sec, it really does not matter much at all, and the compromise between usability, simplicity vs. memory overhead is for me a no-brainer, so I go with usability and simplicity over async features.

That being said, I'm doing a large project with performance in mind, and I have to build it with the private VT API, so I'm kinda forced to rethink a lot of things. You mentioned rate limits in #16 as well as async in this issue. There have also been requests over the years for file extension filters, rotating API keys and much more, and usually the answer have been the same: YOU have to decide on how to do that stuff, this is not an enterprise API with "all you can do" features :)

However, as I'm going to rework a lot of things in the private VT API, async is not a bad idea, as well as simple work queuing (due to potential large amounts of work, and limited batch sizes can be sent to VT private API). So I'm reopening this issue as a reminder.

**Comment by Genbox** [comment]
This has been implemented in 04c3f78710010372bb309dd4e6433515d85f55a3
